cli/docker/container_manager.go

```diff
@@ -0,0 +1,545 @@
+package docker
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bitrise-io/bitrise/log"
+	"github.com/bitrise-io/bitrise/models"
+	"github.com/bitrise-io/go-utils/command"
+	"github.com/bitrise-io/go-utils/v2/redactwriter"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+)
+
+type RunningContainer struct {
+	ID   string
+	Name string
+}
+
+type containerCreateOptions struct {
+	name       string
+	volumes    []string
+	command    string
+	workingDir string
+	user       string
+}
+
+func (rc *RunningContainer) Destroy() error {
+	_, err := command.New("docker", "rm", "--force", rc.Name).RunAndReturnTrimmedCombinedOutput()
+	if err != nil {
+		return fmt.Errorf("remove docker container: %w", err)
+	}
+	return nil
+}
+
+func (rc *RunningContainer) ExecuteCommandArgs(envs []string) []string {
+	args := []string{"exec"}
+
+	for _, env := range envs {
+		args = append(args, "-e", env)
+	}
+
+	args = append(args, rc.Name)
+
+	return args
+}
+
+type ContainerManager struct {
+	logger             DockerLogger
+	workflowContainers map[string]*RunningContainer
+	serviceContainers  map[string][]*RunningContainer
+	client             *client.Client
+
+	mu       sync.Mutex
+	released bool
+}
+
+type DockerLogger struct {
+	logger  log.Logger
+	secrets []string
+}
+
+func (dl *DockerLogger) Infof(format string, args ...interface{}) {
+	log := fmt.Sprintf(format, args...)
+	redacted, _ := dl.Redact(log)
+	dl.logger.Info(redacted)
+}
+
+func (dl *DockerLogger) Errorf(format string, args ...interface{}) {
+	log := fmt.Sprintf(format, args...)
+	redacted, _ := dl.Redact(log)
+	dl.logger.Error(redacted)
+}
+
+func (dl *DockerLogger) Warnf(format string, args ...interface{}) {
+	log := fmt.Sprintf(format, args...)
+	redacted, _ := dl.Redact(log)
+	dl.logger.Warn(redacted)
+}
+
+func (dl *DockerLogger) Redact(s string) (string, error) {
+	src := bytes.NewReader([]byte(s))
+	dstBuf := new(bytes.Buffer)
+	logger := log.NewUtilsLogAdapter()
+	redactWriterDst := redactwriter.New(dl.secrets, dstBuf, &logger)
+
+	if _, err := io.Copy(redactWriterDst, src); err != nil {
+		return "", fmt.Errorf("failed to redact secrets, stream copy failed: %s", err)
+	}
+	if err := redactWriterDst.Close(); err != nil {
+		return "", fmt.Errorf("failed to redact secrets, closing the stream failed: %s", err)
+	}
+
+	redactedValue := dstBuf.String()
+	return redactedValue, nil
+}
+
+func NewContainerManager(logger log.Logger, secrets []string) *ContainerManager {
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		logger.Warnf("Docker client failed to initialize (possibly running on unsupported stack): %s", err)
```

Comment:
nit: in the context of the CLI, there is no stack. I'd rather say unsupported environment. But in the local usecase, IMHO the most common cause of this error is that the Docker daemon is not running.